tensorflow-ops/src/TensorFlow/Minimize.hs

```diff
@@ -0,0 +1,112 @@
+-- Copyright 2016 TensorFlow authors.
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module TensorFlow.Minimize
+    ( Minimizer
+    , minimizeWith
+    , gradientDescent
+    , AdamConfig(..)
+    , adam
+    , adam'
+    ) where
+
+import Control.Monad (zipWithM)
+import Data.Default (Default(..))
+import Data.List (zipWith4)
+import Data.Maybe (fromMaybe)
+
+import qualified TensorFlow.Core as TF
+import qualified TensorFlow.Gradient as TF
+import qualified TensorFlow.Ops as TF hiding (assign, initializedVariable)
+import qualified TensorFlow.Variable as TF
+
+-- | Functions that minimize a loss w.r.t. a set of 'TF.Variable's.
+--
+-- 'Minimizer's are defined as a function of the gradients instead of
+-- the loss so that users can apply transformations to the gradients.
+type Minimizer a =
+    forall m. TF.MonadBuild m =>
```

Comment:
Clarify in the comment that the returned op runs a single step of the optimization (rather than the full minimization loop)?